Plugins/Extra2DA/Extra2DA.cpp

```diff
@@ -0,0 +1,303 @@
+{
+    ArgumentStack stack;
+
+    int32_t result = 0;
+    Types::ObjectID oidCreature = Services::Events::ExtractArgument<Types::ObjectID>(args);
+    CNWSCreature *pCreature = Globals::AppManager()->m_pServerExoApp->GetCreatureByGameObjectID(oidCreature);
+    if ( pCreature && pCreature->m_bPlayerCharacter )
+    {
+        // find first dragon disciple level where scores diverge
+        const uint8_t dragonLevel = GetClassLevel(*pCreature->m_pStats, CLASS_DRAGON_DISCIPLE);
+        const StatBonuses &sb = g_plugin->m_DragonAbilityBonuses;
+        const StatBonuses sbHistory = GetDragonAbilityHistory(pCreature);
+        uint8_t dragonDivergence = 0;
+        for ( uint8_t level = 1; level <= dragonLevel; level++ )
+        {
+            StatBonus awarded = level < sbHistory.size() ? sbHistory[level] : NO_BONUS;
+            StatBonus expected = level < sb.size() ? sb[level] : NO_BONUS;
+            if ( awarded != expected )
+            {
+                dragonDivergence = level;
+                break;
+            }
+        }
+        // find character level at which diverging dragon disciple level was taken
+        if ( dragonDivergence )
+        {
+            const CExoArrayListTemplatedCNWLevelStatsPtr& levelStats = pCreature->m_pStats->m_lstLevelStats;
+            for ( result = 0; result < levelStats.num; result++ )
+                if ( levelStats.element[result]->m_nClass == CLASS_DRAGON_DISCIPLE )
+                    if ( --dragonDivergence == 0 )
+                        break;
+            result++; // 0-based indexing vs 1-based levels
+        }
+    }
+    Services::Events::InsertArgument(stack, result);
+    return stack;
+}
+
+void Extra2DA::LoadDragonAbilityBonuses()
+{
+    // load new bonuses from 2da file
+    C2DA dragon2DA("dragon", false);
+    if ( dragon2DA.Load2DArray() )
+    {
+        if ( dragon2DA.m_nNumRows )
+        {
+            StatBonuses &sb = g_plugin->m_DragonAbilityBonuses;
+            sb.resize(dragon2DA.m_nNumRows, NO_BONUS);
+            for ( uint8_t level = 1; level < dragon2DA.m_nNumRows; level++ )
+                for ( uint8_t stat = 0; stat < 6; stat++ )
+                    if ( !dragon2DA.GetINTEntry(level, stat, &sb[level][stat]) )
+                        LOG_WARNING("Failed to get entry for row %d column %d in dragon.2da.", level, stat);
+            // subtract original bonuses
+            std::vector<uint8_t> statBonusLevels[6] = { {2,4,10,10}, {}, {7}, {9}, {}, {10} };
+            for ( uint8_t stat = 0; stat < 6; stat++ )
+                for ( uint8_t level : statBonusLevels[stat] )
+                    if ( level < sb.size() )
+                        sb[level][stat] -= 2;
+            LOG_DEBUG("Loaded dragon disciple stat bonuses.");
+        }
+        else
+            LOG_WARNING("dragon.2da is empty");
+    }
+    else
+        LOG_DEBUG("dragon.2da not found");
+}
+
+StatBonuses Extra2DA::GetDragonAbilityHistory(CNWSCreature *pCreature)
```

Comment:
You could add nwscript functions for this one and the setter, so users can can save them to a DB etc.

Reply:
Great idea - so obvious now that you mention it.